### PR TITLE
feat(plasma-tokens): Add generate shadow tokens for React-Native

### PR DESCRIPTION
**Pull request by neretin-trike**
:baby_chick: Download canary assets:

[colors--canary.288.4053698929.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.288.4053698929.xml)  
[color_metro_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.288.4053698929.swift)  
[color_metro_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.288.4053698929.kt)  
[color_metro_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.288.4053698929.ts)  
[color_sbermarket_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.288.4053698929.swift)  
[color_sbermarket_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.288.4053698929.kt)  
[color_sbermarket_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.288.4053698929.ts)  
[color_sberprime_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.288.4053698929.swift)  
[color_sberprime_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.288.4053698929.kt)  
[color_sberprime_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.288.4053698929.ts)  
[color_selgros_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.288.4053698929.swift)  
[color_selgros_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.288.4053698929.kt)  
[color_selgros_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.288.4053698929.ts)  
[color_smbusiness_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.288.4053698929.swift)  
[color_smbusiness_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.288.4053698929.kt)  
[color_smbusiness_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.288.4053698929.ts)  
[PlasmaTokensColor--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.288.4053698929.swift)  
[shadow_sbermarket_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.288.4053698929.ts)  
[typo_mage_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.288.4053698929.swift)  
[typo_mage_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.288.4053698929.kt)  
[typo_mage_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.288.4053698929.ts)  
[typo_plasma_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.288.4053698929.swift)  
[typo_plasma_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.288.4053698929.kt)  
[typo_plasma_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.288.4053698929.ts)  
[typo_ruler_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.288.4053698929.swift)  
[typo_ruler_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.288.4053698929.kt)  
[typo_ruler_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.288.4053698929.ts)  
[typo_sage_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.288.4053698929.swift)  
[typo_sage_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.288.4053698929.kt)  
[typo_sage_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.288.4053698929.ts)  
[typo_sbermarket_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.288.4053698929.swift)  
[typo_sbermarket_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.288.4053698929.kt)  
[typo_sbermarket_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.288.4053698929.ts)  
[typo_soulmate_ios-swift--canary.288.4053698929.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.288.4053698929.swift)  
[typo_soulmate_kotlin--canary.288.4053698929.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.288.4053698929.kt)  
[typo_soulmate_react-native--canary.288.4053698929.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.288.4053698929.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.130.0-canary.288.4053698929.0
  npm install @salutejs/plasma-core@1.91.0-canary.288.4053698929.0
  npm install @salutejs/plasma-icons@1.119.0-canary.288.4053698929.0
  npm install @salutejs/plasma-temple@1.127.0-canary.288.4053698929.0
  npm install @salutejs/plasma-tokens@1.33.0-canary.288.4053698929.0
  npm install @salutejs/plasma-ui@1.160.0-canary.288.4053698929.0
  npm install @salutejs/plasma-web@1.157.0-canary.288.4053698929.0
  npm install @salutejs/plasma-cy-utils@0.40.0-canary.288.4053698929.0
  npm install @salutejs/plasma-sb-utils@0.89.0-canary.288.4053698929.0
  npm install @salutejs/plasma-tokens-android@2.40.0-canary.288.4053698929.0
  npm install @salutejs/plasma-tokens-ios-swift@2.40.0-canary.288.4053698929.0
  # or 
  yarn add @salutejs/plasma-b2c@1.130.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-core@1.91.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-icons@1.119.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-temple@1.127.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-tokens@1.33.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-ui@1.160.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-web@1.157.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-cy-utils@0.40.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-sb-utils@0.89.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-tokens-android@2.40.0-canary.288.4053698929.0
  yarn add @salutejs/plasma-tokens-ios-swift@2.40.0-canary.288.4053698929.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-288/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-288/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-288/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-288/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-288/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-288/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-288/temple-storybook/

File: packages/plasma-tokens/data/shadows/base.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                    },
+                    {
+                        color: 'rgba(8, 8, 8, 0.1)',
+                        blurRadius: '12',
+                        spreadRadius: '-2px',
+                        offset: {
+                            x: '0px',
+                            y: '6px',
+                        },
+                    },
```
Comment (Yeti-or) [inline packages/plasma-tokens/data/shadows/base.ts:49]: это норм что в l/m одна тень а в s две ?

File: packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+export const shadowReactNativeCustomFormatter = ({ dictionary }: { dictionary: Dictionary }) => {
+    const sizes = getArray(dictionary, 'subitem');
+    const types = getArray(dictionary, 'item');
+
+    const shadowEnums = getShadowEnumValues(sizes);
+    const shadowsSize = getShadowSizes(sizes);
+    const shadowTokens = getShadowTokens(types, sizes, dictionary.allTokens);
+
+    return getReactNativeTemplate(shadowEnums, shadowsSize, shadowTokens);
+};
```
Comment (Yeti-or) [inline packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts:101]: было бы классно подумать про какие-то тесты на это всё, но хз

**Review by Yeti-or — APPROVED** [review]

File: packages/plasma-tokens/data/shadows/base.ts (hunk 1 above)
Comment (neretin-trike) [reply packages/plasma-tokens/data/shadows/base.ts:49]: ага, так по макетам

File: packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts (hunk 2 above)
Comment (neretin-trike) [reply packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts:101]: ну по-хорошему, как вариант, можно снашпоты кода накрутить как-нибудь

File: packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            const tokens = getShadowTokensByType(sizes, tokenItems, type);
+
+            return `${upperFirstLetter(type)}: {
+    ${tokens}
+  },`;
+        })
+        .join('\n  ');
+
+const getArray = (dictionary: Dictionary, token: string) =>
+    Array.from(new Set(dictionary.allTokens.map((item) => item.attributes?.[token])));
```
Comment (Yakutoc) [inline packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts:90]: @neretin-trike А что будет в результирующем массиве если **attributes** будет undefined?

Comment (neretin-trike) [reply packages/plasma-tokens/lib/generator/formatters/shadowReactNative.ts:90]: а он не будет undefined, т.к. в теме туда в любом случае прокидываются значения. Просто по типу это может быть, но по факту - нет

**Review by Yakutoc — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-288/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-288/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Pull request merged**